File: scratch/scrape.mjs
import fs from 'fs';
import path from 'path';
import { JSDOM } from 'jsdom';
import TurndownService from 'turndown';

const urls = [
  'https://olladns.com/blog/dns-over-https-doh-complete-guide/',
  'https://olladns.com/blog/dns-filtering-explained/',
  'https://olladns.com/blog/dns-firewall-explained-how-dns-firewalls-protect-networks/',
  'https://olladns.com/blog/what-is-dns-how-domain-name-system-works/',
  'https://olladns.com/blog/what-is-dns-security-a-complete-guide-to-protecting-your-network-in-2026/'
];

const outDir = path.resolve('src/content/blog');
const turndownService = new TurndownService({ headingStyle: 'atx', codeBlockStyle: 'fenced' });

async function run() {
  for (const url of urls) {
    try {
      console.log('Fetching ' + url);
      const res = await fetch(url);
      const html = await res.text();
      const dom = new JSDOM(html);
      const document = dom.window.document;

      const title = document.querySelector('meta[property="og:title"]')?.content?.split('·')[0].trim() || document.title;
      const description = document.querySelector('meta[property="og:description"]')?.content || '';
      const pubDate = document.querySelector('meta[property="article:published_time"]')?.content || new Date().toISOString();
      const author = document.querySelector('meta[property="article:author"]')?.content || 'OllaDNS Team';
      const section = document.querySelector('meta[property="article:section"]')?.content || 'Guide';

      const slug = url.split('/').filter(Boolean).pop();

      const articleNode = document.querySelector('article.post-body');
      if (!articleNode) {
        console.log('Could not find article.post-body for ' + url);
        continue;
      }

      // Remove unwanted elements
      const toc = articleNode.querySelector('.post-toc');
      if (toc) toc.remove();

      let markdown = turndownService.turndown(articleNode.innerHTML);

      const frontmatter = `---
title: "${title.replace(/"/g, '\\"')}"
description: "${description.replace(/"/g, '\\"')}"
pubDate: ${new Date(pubDate).toISOString()}
author: "${author}"
tags: ["${section}"]
---
`;

      const finalContent = frontmatter + '\n' + markdown;
      fs.writeFileSync(path.join(outDir, `${slug}.md`), finalContent);
      console.log('Saved ' + slug);
    } catch (err) {
      console.error('Failed on ' + url, err);
    }
  }
}

run();

File: scratch/scrape.py
import urllib.request
import re
import os
from bs4 import BeautifulSoup

urls = [
    'https://olladns.com/blog/dns-over-https-doh-complete-guide/',
    'https://olladns.com/blog/dns-filtering-explained/',
    'https://olladns.com/blog/dns-firewall-explained-how-dns-firewalls-protect-networks/',
    'https://olladns.com/blog/what-is-dns-how-domain-name-system-works/',
    'https://olladns.com/blog/what-is-dns-security-a-complete-guide-to-protecting-your-network-in-2026/'
]

out_dir = 'src/content/blog'

for url in urls:
    try:
        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
        html = urllib.request.urlopen(req).read().decode('utf-8')
        soup = BeautifulSoup(html, 'html.parser')
        
        # Extract title
        og_title = soup.find('meta', property='og:title')
        title = og_title['content'].split('·')[0].strip() if og_title else soup.title.string.strip()
        
        # Extract description
        og_desc = soup.find('meta', property='og:description')
        desc = og_desc['content'] if og_desc else ''
        
        # Extract date
        og_date = soup.find('meta', property='article:published_time')
        pub_date = og_date['content'] if og_date else '2026-08-01'
        
        # Extract section
        og_section = soup.find('meta', property='article:section')
        section = og_section['content'] if og_section else 'Guide'
        
        # Extract body
        body = soup.find('article', class_='post-body')
        if not body:
            print(f"Skipping {url}: No post-body")
            continue
            
        # Remove toc
        toc = body.find(class_='post-toc')
        if toc:
            toc.decompose()
            
        content = str(body)
        
        # Generate markdown frontmatter
        md = f"""---
title: "{title.replace('"', '\\"')}"
description: "{desc.replace('"', '\\"')}"
pubDate: {pub_date}
tags: ["{section}"]
---

{content}
"""
        slug = url.strip('/').split('/')[-1]
        
        filepath = os.path.join(out_dir, f"{slug}.md")
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(md)
            
        print(f"Saved {slug}")
        
    except Exception as e:
        print(f"Error on {url}: {e}")
